subsys: power: define pm policy and power states.

wentongwu opened the pull request:
Define pm policy and power states.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

pabigot commented on include/power/power_state.h:
```diff
@@ -0,0 +1,32 @@
+#define PM_STATE_SUSPEND_TO_DISK BIT(5)
+#define PM_STATE_SOFT_OFF        BIT(6)
+
+#define PM_STATE_BIT_MASK        0x7F
```
`BIT_MASK(7)` would be more clearly tied to the number of states identified by `BIT(n)` above.

pabigot commented on subsys/power/include/pm_policy.h:
```diff
@@ -0,0 +1,182 @@
+ * @brief Set constraint for the power states of the given policy.
+ *
+ * @param policy Pointer to pm_policy instance.
+ * @param states Power states on which set constraint.
```
What is a constraint?

If a state is selected in the bitset is it a state that can be entered, or a state that cannot be entered?

Maybe this should be:

> @param states A bit set identifying PM states that the power management policy is not permitted to select.

wentongwu replied:
@pabigot a state that cannot be entered. Will update the comments much clear. Thanks!

pabigot commented on subsys/power/include/pm_policy.h:
```diff
@@ -0,0 +1,182 @@
+ * @brief Release constraint for the power states of the given policy.
+ *
+ * @param policy Pointer to pm_policy instance.
+ * @param states Power states on which release constraint.
```
Maybe:
> @param states A bit set identifying PM states for which a previous constraint should be cleared, allowing the state to be selected by the policy.

(What happens if two components set a constraint on a particular state?  Is that constraint released if one of them subsequently releases it, but the other doesn't?)

wentongwu replied:
@pabigot it can only be selected again by policy if both components release the constraints they have set. I will update the doc clearly. Thanks

pabigot commented on include/power/power_state.h:
```diff
@@ -13,18 +13,153 @@
+ * Runtime idle is a system sleep state in which all of the cores
+ * enter deepest possible idle state and wait for interrupts, no
+ * requirements for the devices, leaving them at the states where
+ * they are.
```
Could this be:

> Runtime idle is the state entered when the only work to be done is the kernel idle thread.  That should be a system sleep state in which ....

?  I'm not sure that's entirely correct, as there are a very small number of places in the code where the current thread tries to go into idle mode without being a reschedule point, but I think it's worth describing this in terms of the expected behavior of the idle thread.

wentongwu replied:
> as there are a very small number of places in the code where the current thread tries to go into idle mode without being a reschedule point

yes, I think we can enter that mode even not in idle thread.

pabigot commented on include/power/power_state.h:
```diff
@@ -13,18 +13,153 @@
+#define PM_STATE_SUSPEND_TO_RAM        BIT(4)
+
+/**
+ * @brief Suspend to ram state
```
s/ram/disk/

wentongwu replied:
ok

pabigot commented on include/power/power_state.h:
```diff
@@ -13,18 +13,153 @@
+ * This state offers significant energy savings by powering off as much
+ * of the system as possible, where memory should be placed into the
+ * self-refresh mode to retain its contents. The state of devices and
+ * CPUs is saved and held in memory, and it may require some boot-
```
How is the state of the CPU saved?  Is this something that is supported by multiple platforms?

Generally, which of these modes are expected to be ones where the application will restart at the kernel boot and proceed through to main (unless something system-specific intervenes and allows it to jump back to a particular point in the code in an existing thread).

wentongwu replied:
multiple platforms can do this.

pabigot commented on include/power/power_state.h:
```diff
@@ -13,18 +13,153 @@
+ *
+ * This state offers significant energy savings by powering off as much
+ * of the system as possible, including the memory. The contents of
+ * memory are written to disk/flash, and on resume it's read back into
```
"...are written to disk ~/flash~ *or other non-volatile storage*,  and on resume...."

wentongwu replied:
@pabigot Thanks, will update.

pabigot commented on include/power/power_state.h:
```diff
@@ -13,18 +13,153 @@
+ * This state consumes a minimal amount of power and requires a large
+ * latency in order to return to runtime active state. The contents of
+ * system(CPU and memory) will not be preserved, so the system will be
+ * restarted if woken by any wakeup-source.
```
Maybe here is where we first get:

> system will be restarted as if from initial power-up and kernel boot.

wentongwu replied:
@pabigot ok.

pabigot commented on include/power/power_state.h:
```diff
@@ -13,18 +13,153 @@
+/**
+ * @brief PM state bit mask
+ */
+#define PM_STATE_BIT_MASK              BIT_MASK(7)
```
Ah.  Please squash the commits so we don't see the intermediate (pre-documented) state.

wentongwu replied:
yes, Thanks

pabigot commented on subsys/power/include/pm_policy.h:
```diff
@@ -0,0 +1,182 @@
+	 * status and the power management policy where this
+	 * api resides.
+	 */
+	pm_state_t (*next_state)(struct pm_policy *policy);
```
Is there a way to communicate the time of the next known event to this, as currently done with `_sys_suspend()`?

wentongwu replied:
@pabigot if the policy need timeout value to make decision, just use timeout API directly.

vanti commented:
It may help to map these states on real platforms to see how well they apply based on the latest definitions in the code. The TI CC1352 for example supports the following low power states:

IDLE: Turns off the CPU power domain, beyond standard clock-gating (which I presume would correspond to PM_STATE_RUNTIME_IDLE). Minimal latency. Leaves devices on.

STANDBY: Turns off all power domains with the exception of the RTC, GPIO and some CPU logic that allows it to resume from where it left off in the code before sleep. Some latency (about 1 ms). Turns off all devices given their power domains are off.

SHUTDOWN: Turns off all power domains including the RTC. I/O pins are latched, and CPU can be woken up via a GPIO pin. Resumes from reset vector.

I would map them as follow:

```
IDLE = PM_STATE_STANDBY/PM_STATE_SUSPEND_TO_IDLE
STANDBY = PM_STATE_SUSPEND_TO_RAM
SHUTDOWN: PM_STATE_SOFT_OFF
```

I think SHUTDOWN maps well. STANDBY maps ok, even though the CPU doesn't explicitly save its state into RAM, but rather rely on the hardware to retain its state. Arguably we could still claim that is "memory". The main question mark is with IDLE, which does turn off the CPU power domain and leave the devices active, contrary to the definition of PM_STATE_SUSPEND_TO_IDLE. Can we relax its definition to be more inclusive:

```
"The system goes through a normal platform suspend where it puts
all of the cores in deepest possible idle state, and may put peripherals
into low-power states. No operating state is lost (ie. the cpu core
does not lose execution context), so the system can
go back to where it left off easily enough."
```

wentongwu commented:
> The main question mark is with IDLE, which does turn off the CPU power domain and leave the devices active, contrary to the definition of PM_STATE_SUSPEND_TO_IDLE. Can we relax its definition to be more inclusive:
> 
> ```
> "The system goes through a normal platform suspend where it puts
> all of the cores in deepest possible idle state, and may put peripherals
> into low-power states. No operating state is lost (ie. the cpu core
> does not lose execution context), so the system can
> go back to where it left off easily enough."
> ```

done, thanks

wentongwu committed:
subsys: power: define system power state and device power state

Define system power state and device power state.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
subsys: power: define platform pm

Define platform power management.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
subsys: power: define pm policy

Define power management policy.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
device: remove device busy status indication

Remove device busy status indication because it will not impact the
policy decision and it will be covered by power management policy
constraint.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
drivers: change device power transition type

change device power transition callback function type.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu commented:
@pabigot would you like take a look 5f95035db5e5fada6859c28a720dce5edca13a9a?

pabigot commented:
> @pabigot would you like take a look [5f95035](https://github.com/zephyrproject-rtos/zephyr/commit/5f95035db5e5fada6859c28a720dce5edca13a9a)?

I don't like `DEVICE_FOREACH()` as public API because I don't see that type of loop being common, and it certainly isn't something applications should use.

I'm skeptical of the `atomic_t` state in `device_pm`; without seeing the implementation I don't think those eliminate the need for a lock.  Maybe they eliminate a need for a memory barrier somewhere.

I do see the API remains fundamentally asynchronous, thanks.

I have a few other concerns (why is the device power state pointer typeless) but without seeing how it's supposed to work in practice I can't say much.

vanti commented:
> > The main question mark is with IDLE, which does turn off the CPU power domain and leave the devices active, contrary to the definition of PM_STATE_SUSPEND_TO_IDLE. Can we relax its definition to be more inclusive:
> > ```
> > "The system goes through a normal platform suspend where it puts
> > all of the cores in deepest possible idle state, and may put peripherals
> > into low-power states. No operating state is lost (ie. the cpu core
> > does not lose execution context), so the system can
> > go back to where it left off easily enough."
> > ```
> 
> done, thanks

I don't see the change in `power_state.h` for some reason - the word "may" is still left out, which would be key for IDLE to be mapped to PM_STATE_SUSPEND_TO_IDLE.

wentongwu committed:
device: device power management API

Define device power management API.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

vanti commented on include/power/power_state.h:
```diff
@@ -0,0 +1,158 @@
+ * @brief Suspend to idle state
+ *
+ * The system goes through a normal platform suspend where it puts
+ * all of the cores in deepest possible idle state and puts peripherals
```
Just to reiterate and clarify my previous comment, this sentence should be changed to "idle state and may put peripherals into low-power states" in order to match state S1 in ACPI, which says "Devices that do not indicate they must remain on **may** be powered off" (from https://en.wikipedia.org/wiki/Advanced_Configuration_and_Power_Interface#Global_states). Powering off the peripherals should be optional.

wentongwu committed:
add more for the APIs.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
the code I have.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>
